server/session.go

```diff
@@ -168,16 +173,31 @@ func (p *Plugin) addUserSession(state *callState, callsEnabled *bool, userID, co
 		UserID: userID,
 		JoinAt: time.Now().UnixMilli(),
 	}
+	defer func() {
+		// We need to make sure and keep the state consistent in case of error since it can be shared
+		// with other operations in the same batch.
+		if retErr != nil {
+			// Undo in case of error
+			delete(state.sessions, connID)
+		}
+	}()
 
 	if newHostID := state.getHostID(p.getBotID()); newHostID != state.Call.GetHostID() {
 		state.Call.Props.Hosts = []string{newHostID}
-		p.publishWebSocketEvent(wsEventCallHostChanged, map[string]interface{}{
-			"hostID": newHostID,
-		}, &WebSocketBroadcast{
-			ChannelID:           channelID,
-			ReliableClusterSend: true,
-			UserIDs:             getUserIDsFromSessions(state.sessions),
-		})
+		defer func() {
+			if retErr == nil {
+				p.publishWebSocketEvent(wsEventCallHostChanged, map[string]interface{}{
+					"hostID": newHostID,
+				}, &WebSocketBroadcast{
+					ChannelID:           channelID,
+					ReliableClusterSend: true,
+					UserIDs:             getUserIDsFromSessions(state.sessions),
+				})
+			} else {
+				// Undo in case of error
```

Comment:
So this means we have to hang on to err and undo all state changes now? 
It seems like adding yet another complication... Is there any way around it? Maybe a "change state or rollback" to make it simpler?

Reply:
I know but the simpler alternative would be to deep clone the whole state and return the original in case of failure. Do we really want to do that every time just to cover an error case that shouldn't happen under normal circumstances? Happy discuss different solutions of course, just couldn't think of anything better.

Reply:
It's up to you of course, and I'm not sure of the feasibility and cost of doing that. But I guess I would say as devil's advocate: consider a) the mental cost of having to double check if this is a place where we need to catch errors and revert changes + do that manually, vs b) the cost of auto clone + reverting.

Reply:
It's a fair point. I'll run some benchmarks as there's a good chance cloning performance to be a blip in the whole context of why we are doing this.

Reply:
@cpoile Take a look. I made the change as state cloning performance shouldn't be a concern.

```
goos: linux
goarch: amd64
pkg: github.com/mattermost/mattermost-plugin-calls/server
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkCallStateClone
BenchmarkCallStateClone/10_sessions
BenchmarkCallStateClone/10_sessions-8         	  568357	      1998 ns/op	    2020 B/op	      16 allocs/op
BenchmarkCallStateClone/100_sessions
BenchmarkCallStateClone/100_sessions-8        	   77806	     15612 ns/op	   17016 B/op	     107 allocs/op
BenchmarkCallStateClone/1000_sessions
BenchmarkCallStateClone/1000_sessions-8       	    8863	    127430 ns/op	  138168 B/op	    1007 allocs/op
BenchmarkCallStateClone/10000_sessions
BenchmarkCallStateClone/10000_sessions-8      	     800	   1358737 ns/op	 1259576 B/op	   10007 allocs/op
BenchmarkCallStateClone/0_sessions
BenchmarkCallStateClone/0_sessions-8          	 2141230	       586.3 ns/op	     800 B/op	       5 allocs/op
PASS
ok  	github.com/mattermost/mattermost-plugin-calls/server	73.514s
```